private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -544,49 +474,129 @@ func (t *transitiveClosure) addElement(
-			if !ok {
-				return fmt.Errorf("missing %q", typeName)
-			}
-			err := t.addElement(typeDescriptor, descriptorInfo.file, imageIndex, opts)
-			if err != nil {
-				return err
-			}
-		case descriptorpb.FieldDescriptorProto_TYPE_DOUBLE,
-			descriptorpb.FieldDescriptorProto_TYPE_FLOAT,
-			descriptorpb.FieldDescriptorProto_TYPE_INT64,
-			descriptorpb.FieldDescriptorProto_TYPE_UINT64,
-			descriptorpb.FieldDescriptorProto_TYPE_INT32,
-			descriptorpb.FieldDescriptorProto_TYPE_FIXED64,
-			descriptorpb.FieldDescriptorProto_TYPE_FIXED32,
-			descriptorpb.FieldDescriptorProto_TYPE_BOOL,
-			descriptorpb.FieldDescriptorProto_TYPE_STRING,
-			descriptorpb.FieldDescriptorProto_TYPE_BYTES,
-			descriptorpb.FieldDescriptorProto_TYPE_UINT32,
-			descriptorpb.FieldDescriptorProto_TYPE_SFIXED32,
-			descriptorpb.FieldDescriptorProto_TYPE_SFIXED64,
-			descriptorpb.FieldDescriptorProto_TYPE_SINT32,
-			descriptorpb.FieldDescriptorProto_TYPE_SINT64:
-			// nothing to follow, custom options handled below.
-		default:
-			return fmt.Errorf("unknown field type %d", typedDescriptor.GetType())
-		}
-		if err := t.exploreCustomOptions(typedDescriptor, descriptorInfo.file, imageIndex, opts); err != nil {
+	default:
+		return errorUnsupportedFilterType(descriptor, descriptorInfo.fullName)
+	}
+
+	return nil
+}
+
+func errorUnsupportedFilterType(descriptor namedDescriptor, fullName string) error {
+	var descriptorType string
+	switch d := descriptor.(type) {
+	case *descriptorpb.FileDescriptorProto:
+		descriptorType = "file"
+	case *descriptorpb.DescriptorProto:
+		descriptorType = "message"
+	case *descriptorpb.FieldDescriptorProto:
+		if d.Extendee != nil {
+			descriptorType = "extension field"
+		} else {
+			descriptorType = "non-extension field"
+		}
+	case *descriptorpb.OneofDescriptorProto:
+		descriptorType = "oneof"
+	case *descriptorpb.EnumDescriptorProto:
+		descriptorType = "enum"
+	case *descriptorpb.EnumValueDescriptorProto:
+		descriptorType = "enum value"
+	case *descriptorpb.ServiceDescriptorProto:
+		descriptorType = "service"
+	case *descriptorpb.MethodDescriptorProto:
+		descriptorType = "method"
+	default:
+		descriptorType = fmt.Sprintf("%T", d)
+	}
+	return fmt.Errorf("%s is unsupported filter type: %s", fullName, descriptorType)
+}
+
+func (t *transitiveClosure) addEnclosing(descriptor namedDescriptor, enclosingFile string, imageIndex *imageIndex, opts *imageFilterOptions) error {
+	for descriptor != nil {
+		_, isMsg := descriptor.(*descriptorpb.DescriptorProto)
+		_, isSvc := descriptor.(*descriptorpb.ServiceDescriptorProto)
+		if !isMsg && !isSvc {
+			break // not an enclosing type
+		}
+		if _, ok := t.elements[descriptor]; ok {
+			break // already in closure
+		}
+		t.elements[descriptor] = inclusionModeEnclosing
+		if err := t.exploreCustomOptions(descriptor, enclosingFile, imageIndex, opts); err != nil {
+			return err
+		}
+		// now move into this element's parent
+		descriptor = imageIndex.ByDescriptor[descriptor].parent
```

Comment:
I'm having a little trouble following the control flow here. From what I can see we only reach this method through:

```
descriptorInfo := imageIndex.ByDescriptor[descriptor]
...
// if this type is enclosed inside another, add enclosing types
if err := t.addEnclosing(descriptorInfo.parent, descriptorInfo.file, imageIndex, opts); err != nil {
	return err
}
// <no use of descriptorInfo.parent>
```

In the calling function (`addElement`), `descriptorInfo` is a `elementInfo` struct (no pointer).

I commented this out and test started failing so I must be misreading something, but what is going on here is not clear to me.

Reply:
Per F2F, there was confusion around this being a loop vs. a conditional. It is not really intuitive that it's a loop, so it's easy to misread `for` as `if` at the top. Added a comment to improve clarity.